File: mobile/lib/services/license_service.dart
// services/license_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:crypto/crypto.dart';
import 'package:device_info_plus/device_info_plus.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'firebase_service.dart';

const _secret = 'MF-4g9zK2#pL8mXqR5vN1wJ7cT0bY6hD';
const _prefKey = 'mf_license';
const _warnDays = 7;

class LicenseResult {
  final bool ok;
  final String msg;
  final int? daysLeft;
  final bool warn;
  final String? type;
  final String profileId;
  final String key;

  LicenseResult({
    required this.ok,
    required this.msg,
    this.daysLeft,
    this.warn = false,
    this.type,
    this.profileId = '',
    this.key = '',
  });
}

class LicenseService {
  static Future<String> getDeviceId() async {
    final info = DeviceInfoPlugin();
    if (Platform.isAndroid) {
      final android = await info.androidInfo;
      final raw = '${android.model}-${android.id}-${android.serialNumber}';
      return sha256.convert(utf8.encode(raw)).toString().substring(0, 32);
    } else if (Platform.isIOS) {
      final ios = await info.iosInfo;
      final raw = '${ios.name}-${ios.identifierForVendor}';
      return sha256.convert(utf8.encode(raw)).toString().substring(0, 32);
    }
    return 'unknown';
  }

  static String _sign(String key, String deviceId) {
    final payload = '$key:$deviceId';
    final hmac = Hmac(sha256, utf8.encode(_secret));
    return hmac.convert(utf8.encode(payload)).toString();
  }

  static Future<void> _saveLocal(String key, String deviceId, String profileId) async {
    final prefs = await SharedPreferences.getInstance();
    final sig = _sign(key, deviceId);
    await prefs.setString(_prefKey, jsonEncode({
      'key': key,
      'device_id': deviceId,
      'profile_id': profileId,
      'sig': sig,
    }));
  }

  static Future<Map<String, dynamic>?> _loadLocal() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_prefKey);
    if (raw == null) return null;
    try {
      final data = jsonDecode(raw) as Map<String, dynamic>;
      final expected = _sign(data['key'], data['device_id']);
      if (data['sig'] != expected) return null;
      return data;
    } catch (_) {
      return null;
    }
  }

  static Future<String> _getIp() async {
    try {
      final interfaces = await NetworkInterface.list();
      for (final i in interfaces) {
        for (final addr in i.addresses) {
          if (addr.type == InternetAddressType.IPv4 && !addr.isLoopback) {
            return addr.address;
          }
        }
      }
    } catch (_) {}
    return 'desconocida';
  }

  static Future<LicenseResult> activate(String key) async {
    final deviceId = await getDeviceId();

    final lic = await FirebaseService.getLicense(key);
    if (lic == null) return LicenseResult(ok: false, msg: 'Key inválida.');
    if (lic['active'] != true) return LicenseResult(ok: false, msg: 'Key desactivada.');

    // Verificar expiración
    int? daysLeft;
    final expiresAt = lic['expires_at'];
    if (expiresAt != null && expiresAt != 'never') {
      final exp = DateTime.parse(expiresAt).toUtc();
      final now = DateTime.now().toUtc();
      if (now.isAfter(exp)) return LicenseResult(ok: false, msg: 'Key vencida.');
      daysLeft = exp.difference(now).inDays;
    }

    // Verificar plataforma
    final platform = lic['platform'] ?? 'both';
    if (platform == 'pc') {
      return LicenseResult(ok: false, msg: 'Esta key es solo para PC.');
    }

    // Verificar dispositivos
    final devices = await FirebaseService.getDevicesForKey(key);
    final deviceIds = devices.map((d) => d['device_id']).toList();
    final maxDevices = (lic['max_devices'] ?? 1) as int;

    final profileId = (lic['profile_id'] ?? '') as String;

    if (deviceIds.contains(deviceId)) {
      await _saveLocal(key, deviceId, profileId);
      return LicenseResult(ok: true, msg: 'Licencia válida.', daysLeft: daysLeft, type: lic['type'], profileId: profileId, key: key);
    }

    if (devices.length >= maxDevices) {
      return LicenseResult(
        ok: false,
        msg: 'Límite de dispositivos alcanzado ($maxDevices). Contacta al administrador.',
      );
    }

    // Registrar dispositivo
    final info = DeviceInfoPlugin();
    String hostname = 'Android';
    if (Platform.isAndroid) {
      final android = await info.androidInfo;
      hostname = '${android.brand} ${android.model}';
    } else if (Platform.isIOS) {
      final ios = await info.iosInfo;
      hostname = ios.name;
    }

    await FirebaseService.registerDevice({
      'device_id': deviceId,
      'key_id': key,
      'platform': Platform.isIOS ? 'ios' : 'mobile',
      'hostname': hostname,
      'ip': await _getIp(),
      'registered_at': DateTime.now().toUtc().toIso8601String(),
      'last_seen': DateTime.now().toUtc().toIso8601String(),
    });

    await _saveLocal(key, deviceId, profileId);
    return LicenseResult(ok: true, msg: 'Licencia activada.', daysLeft: daysLeft, type: lic['type'], profileId: profileId, key: key);
  }

  static Future<LicenseResult> validate() async {
    final local = await _loadLocal();
    if (local == null) return LicenseResult(ok: false, msg: 'No hay licencia activada.');

    final deviceId = await getDeviceId();
    if (local['device_id'] != deviceId) {
      return LicenseResult(ok: false, msg: 'Licencia no válida para este dispositivo.');
    }

    try {
      final lic = await FirebaseService.getLicense(local['key']);
      if (lic == null) return LicenseResult(ok: false, msg: 'Key no encontrada.');
      if (lic['active'] != true) return LicenseResult(ok: false, msg: 'Licencia revocada.');

      int? daysLeft;
      bool warn = false;
      final expiresAt = lic['expires_at'];
      if (expiresAt != null && expiresAt != 'never') {
        final exp = DateTime.parse(expiresAt).toUtc();
        final now = DateTime.now().toUtc();
        if (now.isAfter(exp)) return LicenseResult(ok: false, msg: 'Licencia vencida. Renueva tu plan.');
        daysLeft = exp.difference(now).inDays;
        warn = daysLeft <= _warnDays;
      }

      await FirebaseService.updateDeviceLastSeen(deviceId, await _getIp());
      return LicenseResult(ok: true, msg: 'Licencia válida.', daysLeft: daysLeft, warn: warn, type: lic['type'], profileId: local['profile_id'] ?? '', key: local['key'] ?? '');
    } catch (e) {
      return LicenseResult(ok: false, msg: 'Error al validar: $e');
    }
  }

  static Future<void> saveSession(Map<String, dynamic> session) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('mf_session', jsonEncode(session));
  }

  static Future<Map<String, dynamic>?> loadSession() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString('mf_session');
    if (raw == null) return null;
    try {
      return jsonDecode(raw) as Map<String, dynamic>;
    } catch (_) {
      return null;
    }
  }

  static Future<void> clearSession() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('mf_session');
  }
}

File: desktop_java/src/main/java/com/mfagent/service/LicenseService.java
package com.mfagent.service;

import com.google.cloud.firestore.Firestore;
import com.google.cloud.firestore.QueryDocumentSnapshot;
import com.google.cloud.firestore.QuerySnapshot;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

/**
 * Validación y activación de licencias contra Firebase.
 * Equivalente a lic_manager/license_manager.py
 *
 * - HMAC-SHA256 para firmar el .license local
 * - Hardware ID vinculado al dispositivo
 * - Soporte multi-dispositivo (max_devices)
 * - Aviso cuando quedan ≤7 días
 */
public class LicenseService {

    private static final byte[] SECRET = "MF-4g9zK2#pL8mXqR5vN1wJ7cT0bY6hD".getBytes();
    private static final int WARN_DAYS = 7;
    private static final Gson GSON = new Gson();

    // ── Hardware ID ───────────────────────────────────────────────────────────

    public String getHardwareId() {
        try {
            String raw = InetAddress.getLocalHost().getHostName()
                    + "-" + System.getProperty("os.arch")
                    + "-" + getMacAddress();
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] hash = md.digest(raw.getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte b : hash) sb.append(String.format("%02x", b));
            return sb.substring(0, 32);
        } catch (Exception e) {
            return "unknown-device";
        }
    }

    private String getMacAddress() {
        try {
            NetworkInterface ni = NetworkInterface.getByInetAddress(InetAddress.getLocalHost());
            if (ni == null) return "00:00:00:00:00:00";
            byte[] mac = ni.getHardwareAddress();
            if (mac == null) return "00:00:00:00:00:00";
            StringBuilder sb = new StringBuilder();
            for (byte b : mac) sb.append(String.format("%02x", b));
            return sb.toString();
        } catch (Exception e) {
            return "00:00:00:00:00:00";
        }
    }

    // ── HMAC firma ────────────────────────────────────────────────────────────

    private String sign(String key, String deviceId) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(SECRET, "HmacSHA256"));
            byte[] hash = mac.doFinal((key + ":" + deviceId).getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte b : hash) sb.append(String.format("%02x", b));
            return sb.toString();
        } catch (Exception e) {
            return "";
        }
    }

    // ── .license local ────────────────────────────────────────────────────────

    private Path licenseFile() {
        return ProfilePaths.licenseFile();
    }

    public void saveLocal(String key, String deviceId, String profileId) {
        Map<String, String> data = new HashMap<>();
        data.put("key",        key);
        data.put("device_id",  deviceId);
        data.put("profile_id", profileId);
        data.put("sig",        sign(key, deviceId));
        try {
            Files.writeString(licenseFile(), GSON.toJson(data));
        } catch (Exception e) {
            System.err.println("[License] Error guardando .license: " + e.getMessage());
        }
    }

    public Map<String, String> loadLocal() {
        try {
            if (!Files.exists(licenseFile())) return Map.of();
            String json = Files.readString(licenseFile());
            Map<String, String> data = GSON.fromJson(json,
                    new TypeToken<Map<String, String>>(){}.getType());
            // Verificar firma
            String expected = sign(data.get("key"), data.get("device_id"));
            if (!expected.equals(data.get("sig"))) return Map.of();
            return data;
        } catch (Exception e) {
            return Map.of();
        }
    }

    public String loadLocalKey() {
        return loadLocal().getOrDefault("key", null);
    }

    public String loadLocalProfileId() {
        return loadLocal().getOrDefault("profile_id", null);
    }

    // ── Activar key ───────────────────────────────────────────────────────────

    public LicenseResult activate(String key) {
        try {
            Firestore db = FirebaseService.getLicDb();
            String deviceId = getHardwareId();

            var licDoc = db.collection("licenses").document(key).get().get();
            if (!licDoc.exists()) return LicenseResult.fail("Key inválida.");

            Map<String, Object> data = licDoc.getData();
            if (!Boolean.TRUE.equals(data.get("active")))
                return LicenseResult.fail("Key desactivada.");

            // Verificar plataforma
            String platform = str(data, "platform");
            if ("mobile".equals(platform))
                return LicenseResult.fail("Esta key es solo para dispositivos móviles.");

            // Verificar expiración
            Integer daysLeft = null;
            String expiresAt = str(data, "expires_at");
            if (expiresAt != null && !expiresAt.isEmpty() && !"never".equals(expiresAt)) {
                ZonedDateTime exp = ZonedDateTime.parse(expiresAt, DateTimeFormatter.ISO_DATE_TIME)
                        .withZoneSameInstant(ZoneOffset.UTC);
                ZonedDateTime now = ZonedDateTime.now(ZoneOffset.UTC);
                if (now.isAfter(exp)) return LicenseResult.fail("Key vencida.");
                daysLeft = (int) ChronoUnit.DAYS.between(now, exp);
            }

            // Verificar dispositivos
            QuerySnapshot devSnap = db.collection("devices")
                    .whereEqualTo("key_id", key).get().get();
            List<QueryDocumentSnapshot> devices = devSnap.getDocuments();
            List<String> deviceIds = devices.stream()
                    .map(d -> { String v = d.getString("device_id"); return v != null ? v : ""; })
                    .toList();

            int maxDevices = ((Number) data.getOrDefault("max_devices", 1)).intValue();

            if (!deviceIds.contains(deviceId) && deviceIds.size() >= maxDevices)
                return LicenseResult.fail("Límite de dispositivos alcanzado (" + maxDevices + ").");

            // Registrar dispositivo si es nuevo
            if (!deviceIds.contains(deviceId)) {
                Map<String, Object> devData = new HashMap<>();
                devData.put("device_id",     deviceId);
                devData.put("key_id",         key);
                devData.put("platform",       "pc");
                devData.put("hostname",       InetAddress.getLocalHost().getHostName());
                devData.put("registered_at",  Instant.now().toString());
                devData.put("last_seen",      Instant.now().toString());
                db.collection("devices").document(deviceId).set(devData).get();
            }

            String profileId = str(data, "profile_id");
            saveLocal(key, deviceId, profileId != null ? profileId : "");
            return LicenseResult.ok(profileId, daysLeft);

        } catch (Exception e) {
            return LicenseResult.fail("Error al activar: " + e.getMessage());
        }
    }

    // ── Validar licencia guardada ─────────────────────────────────────────────

    public LicenseResult validate() {
        Map<String, String> local = loadLocal();
        if (local.isEmpty()) return LicenseResult.fail("No hay licencia activada.");

        String key      = local.get("key");
        String deviceId = local.get("device_id");

        if (!getHardwareId().equals(deviceId))
            return LicenseResult.fail("Licencia no válida para este equipo.");

        try {
            Firestore db = FirebaseService.getLicDb();
            var licDoc = db.collection("licenses").document(key).get().get();
            if (!licDoc.exists()) return LicenseResult.fail("Key no encontrada.");

            Map<String, Object> data = licDoc.getData();
            if (!Boolean.TRUE.equals(data.get("active")))
                return LicenseResult.fail("Licencia revocada.");

            Integer daysLeft = null;
            boolean warn = false;
            String expiresAt = str(data, "expires_at");
            if (expiresAt != null && !expiresAt.isEmpty() && !"never".equals(expiresAt)) {
                ZonedDateTime exp = ZonedDateTime.parse(expiresAt, DateTimeFormatter.ISO_DATE_TIME)
                        .withZoneSameInstant(ZoneOffset.UTC);
                ZonedDateTime now = ZonedDateTime.now(ZoneOffset.UTC);
                if (now.isAfter(exp)) return LicenseResult.fail("Licencia vencida. Renueva tu plan.");
                daysLeft = (int) ChronoUnit.DAYS.between(now, exp);
                warn = daysLeft <= WARN_DAYS;
            }

            // Actualizar last_seen
            db.collection("devices").document(deviceId)
                    .update("last_seen", Instant.now().toString());

            String profileId = local.get("profile_id");
            return LicenseResult.ok(profileId, daysLeft).withWarn(warn);

        } catch (Exception e) {
            return LicenseResult.fail("Error al validar: " + e.getMessage());
        }
    }

    private String str(Map<String, Object> m, String key) {
        Object v = m.get(key); return v != null ? v.toString() : null;
    }

    // ── Resultado ─────────────────────────────────────────────────────────────

    public static class LicenseResult {
        public final boolean ok;
        public final String  message;
        public final String  profileId;
        public final Integer daysLeft;
        public boolean warn;

        private LicenseResult(boolean ok, String message, String profileId, Integer daysLeft) {
            this.ok = ok; this.message = message;
            this.profileId = profileId; this.daysLeft = daysLeft;
        }

        public static LicenseResult ok(String profileId, Integer daysLeft) {
            return new LicenseResult(true, "Licencia válida.", profileId, daysLeft);
        }
        public static LicenseResult fail(String msg) {
            return new LicenseResult(false, msg, null, null);
        }
        public LicenseResult withWarn(boolean w) { this.warn = w; return this; }
    }
}
